### DISCUSSION TITLE
MongoDB 100% CPU issue

### DISCUSSION THREAD

**Discussion by coinofalex** [strapi/strapi#6603] [Legacy Topics]
**Describe the bug**
We have endpoint, when it is called we experience 100% CPU usage on MongoDB instance (it was AWS EC2 t3.large at the time of issue) and failure in the end. When we moved to AWS EC2 t3.xlarge it chewed the query and returned queried data. At that moment we experienced around 90% of CPU usage, which soon will lead to another increase.

~~**Steps to reproduce the behavior**~~ it is impossible, as we fixed that.

Request looks like this:
entitycampaigns?_limit=-1&brands.name=COM&entityprojects.project_id=entity&channels.name=desktop&languages.language_id=en

**Expected behavior**
Mongo query has fast response and consumes much less CPU. Expected to see response from MongoDB in 200ms and CPU usage not more than 1-10% (depending on whether this request is hard for it)

**Screenshots**
![image (19)](https://user-images.githubusercontent.com/66669892/84151571-3e803180-aa6c-11ea-9716-c04158231c24.png)
![image-2020-05-29-19-31-39-028 (1)](https://user-images.githubusercontent.com/66669892/84151580-417b2200-aa6c-11ea-8773-3f6aa5acde85.png)
![image (20)](https://user-images.githubusercontent.com/66669892/84151581-4213b880-aa6c-11ea-9689-326d2804b3b1.png)
![image-2020-05-29-19-32-34-982 (1)](https://user-images.githubusercontent.com/66669892/84151583-42ac4f00-aa6c-11ea-931a-9fb802f2d2f5.png)

**Code snippets**
This is our `/api/entitycampaign/models/Entitycampaign.settings.json`
```
{
  "kind": "collectionType",
  "connection": "default",
  "collectionName": "entitycampaign",
  "info": {
    "name": "entitycampaign",
    "description": ""
  },
  "options": {
    "timestamps": [
      "createdAt",
      "updatedAt"
    ]
  },
  "attributes": {
    "external_id": {
      "type": "string"
    },
    "name": {
      "type": "string"
    },
    "description_top": {
      "type": "text"
    },
    "description": {
      "type": "text"
    },
    "detail_description_name": {
      "type": "text"
    },
    "detail_description_top": {
      "type": "text"
    },
    "detail_description": {
      "type": "text"
    },
    "rules": {
      "type": "text"
    },
    "image": {
      "model": "file",
      "via": "related",
      "plugin": "upload",
      "required": false
    },
    "brands": {
      "via": "entitycampaigns",
      "collection": "brand",
      "dominant": true
    },
    "channels": {
      "collection": "channel",
      "via": "entitycampaigns",
      "dominant": true
    },
    "languages": {
      "collection": "language",
      "via": "entitycampaigns",
      "dominant": true
    },
    "entityprojects": {
      "collection": "entityproject",
      "via": "entitycampaigns",
      "dominant": true
    }
  }
}
```

**System**

- Node.js version: v10.20.1 (not if v12 here will help)
- NPM version: N/A
- Strapi version: v3.0.0-beta.19.3

**Additional info**
Pay attention to that we have different relations most of which are many to many (whether it affects or not, is you to suggest)
Pay attention to limit=-1 param as well you to suggest

Due to our NDA I cannot place some information like urls etc.

**Comment by derrickmehaffy** [comment]
`_limit=-1` your problem is there, that literally dumps the entire contents of the collection and all level 1 related entries.

Suggesting you use proper paging with a normal _limit and using _start.

As a side point: `v3.0.0-beta.19.3` beta versions are no longer supported. I am going to move this to a discussion.

**Reply by derrickmehaffy** [reply]
And yes manyToMany is an extremely taxing relationship type that I generally advise you don't use unless you absolutely need to store the relations on both sides.

I would say this is not inherently a bug, but just a really poorly optimized API call

**Reply by derrickmehaffy** [reply]
And taking a second look you are also using deep filtering which specifically warns you on the console that you really shouldn't rely on that for large queries

**Reply by coinofalex** [reply]
Thanks a lot, this is something we expected, thanks for confirming. Anything else beyond that?

**Reply by derrickmehaffy** [reply]
Nothing I can see off the top of my head but naturally I would need to look at both sides of the code base to really dig in which is not an option due to your NDA.

Best case for you guys would be to build custom queries to handle the data exactly as you need: 
https://strapi.io/documentation/3.0.0-beta.x/concepts/queries.html#mongoose

**Reply by derrickmehaffy** [reply]
(As a small suggestion, I would highly suggest you guys look at migrating to the stable release also: https://strapi.io/documentation/v3.x/migration-guide/migration-guide-beta.20-to-3.0.0.html )
